File: api/app/schemas/returns.py
"""Returns: request, list, and the HITL (human-in-the-loop) path.

within-window logic is computed server-side from the order/policy; a request
outside the window yields ``return_window_closed`` (409) on the direct path, but the
agent-assisted path instead creates a return in ``hitl_pending`` for a human to
approve/reject (the J-BUY-06 / J-SUP-03 HITL flow).
"""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.schemas.enums import ReturnReason, ReturnStatus
from app.schemas.envelope import CamelModel


class ReturnItemRequest(CamelModel):
    """One line of a return request (references an order item)."""

    order_item_id: str
    qty: int = Field(ge=1, le=999)


class ReturnCreate(CamelModel):
    """POST /orders/{order_id}/returns — request a return."""

    reason_code: ReturnReason
    note: str | None = Field(default=None, max_length=2000)
    items: list[ReturnItemRequest] = Field(min_length=1)


class ReturnItemOut(CamelModel):
    """A persisted return line."""

    id: str
    order_item_id: str
    qty: int = Field(ge=1)


class ReturnOut(CamelModel):
    """A return with status (incl. ``hitl_pending`` for the HITL path)."""

    id: str
    order_id: str
    status: ReturnStatus
    reason_code: ReturnReason
    note: str | None = None
    within_window: bool
    approved_by: str | None = None
    items: list[ReturnItemOut]
    created_at: datetime
    resolved_at: datetime | None = None


class ReturnDecision(CamelModel):
    """PATCH /returns/{id} — support/admin (or HITL) approve/reject/refund."""

    status: ReturnStatus = Field(
        description="Target status: approved | rejected | refunded.",
    )
    note: str | None = Field(default=None, max_length=2000)

File: web/src/lib/mock/returns.ts
import type { ReturnCreate, ReturnItemOut, ReturnOut } from "@/lib/api-types";
import { getOrder, markReturnRequested } from "@/lib/mock/orders";

/**
 * MOCK returns store (ADR-0037). Backs `POST /orders/{id}/returns` (+ a cheap
 * list) deterministically. The HITL path is modeled off the order's window:
 *  - order within the return window (`return_eligible`) → a `requested` return
 *    (the normal path; a human/support still decides, but it's in-window).
 *  - order OUTSIDE the window → the return lands in `hitl_pending` — the
 *    contract's agent-assisted HITL deferral (`order-status-hitl-pending`),
 *    surfaced so the human-in-the-loop decision is visible in the UI.
 */

let returnSeq = 5000;
const returns: ReturnOut[] = [];

export function createReturn(
  orderId: string,
  body: ReturnCreate,
): ReturnOut | { error: "not_found" } {
  const order = getOrder(orderId);
  if (!order) return { error: "not_found" };

  const withinWindow = order.return_eligible === true;
  const items: ReturnItemOut[] = body.items.map((it, i) => ({
    id: `ri_${++returnSeq}_${i}`,
    order_item_id: it.order_item_id,
    qty: it.qty,
  }));

  const now = new Date().toISOString();
  const ret: ReturnOut = {
    id: `ret_${++returnSeq}`,
    order_id: orderId,
    // Out-of-window → HITL deferral; in-window → standard request.
    status: withinWindow ? "requested" : "hitl_pending",
    reason_code: body.reason_code,
    note: body.note ?? null,
    within_window: withinWindow,
    approved_by: null,
    items,
    created_at: now,
    resolved_at: null,
  };
  returns.push(ret);
  markReturnRequested(orderId);
  return ret;
}

export function listReturns(orderId?: string): ReturnOut[] {
  return returns.filter((r) => !orderId || r.order_id === orderId);
}
